chore(deps): Support Python 3.13

reneleonhardt opened the pull request:
Closes #3037

reneleonhardt committed:
chore(deps): Support Python 3.13 and drop 3.9

Closes #3037

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Aider-AI/aider?pullRequest=3756) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Aider-AI/aider?pullRequest=3756) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Aider-AI/aider?pullRequest=3756) it.</sub>

reneleonhardt commented:
If the tests are green, please try yourself @paul-gauthier 😄

reneleonhardt committed:
Workaround uv bug to allow more updates

reneleonhardt committed:
Add Python 3.10 requirements

reneleonhardt commented:
FYI there were many llama-index-core releases since you pinned 0.12.26 3 weeks ago, maybe test_help will work again 🤞😅

claui commented on requirements/requirements-help.in:
```diff
@@ -1,11 +1,11 @@
+numpy>=2
 
-# Mac x86 only supports 2.2.2
+# Drop Mac x86 support
```
See also https://github.com/Aider-AI/aider/pull/3444#issuecomment-2702280870 for an ongoing discussion about keeping or dropping Mac x86 support.

reneleonhardt replied:
Sure, but you can't fight time 😅 torch 2.7 just came out, they dropped Mac x86 support a year ago.
If it's any relief, Apple stopped selling Intel Macs 2 years ago and will probably stop supporting them in macOS 17 next year.

claui replied:
I think that this decision should be ultimately up to the project maintainer. /cc @paul-gauthier

reneleonhardt committed:
More updates

reneleonhardt commented:
@claui How is the review progressing?

claui commented:
@reneleonhardt Sorry but I’m not actively reviewing this PR. The intent of my comment was to make sure you’re aware of [the existing discussion](https://github.com/Aider-AI/aider/pull/3444#issuecomment-2702280870). If GitHub shows me as a reviewer for this PR, then that’s misleading. Sorry about that.

In fact, I’ve filed #3516 (motivated by the Intel Mac requirement) back in March, i.e. before you filed this competing PR. I’m still waiting on @paul-gauthier’s feedback on it.

reneleonhardt commented:
Yeah, it's always confusing to me that a contributor isn't allowed to request a review 😞
You're right, it took me a long time to work around all those requirement lockfiles, real dependency management with uv or poetry would be so much easier 👍

reneleonhardt commented:
So... AI help for faster contributions isn't wanted and human help for dropping Python 3.9 isn't wanted either, now my work will never get reviewed 👍

ei-grad commented:
@reneleonhardt Many thanks for this PR! I've been using your PR locally for several weeks, and it worked perfectly both in 3.12/3.13 envs – truly appreciate you sharing this.

It looks like Python 3.13 support might be getting integrated by the maintainer more gradually, in separate parts – perhaps that's why this PR has some conflicts now. That's a challenging situation, and I can imagine it might be disheartening.

Despite that, it really shows that any good contribution you share can make a real difference for someone – your work here has definitely made one for me. Hope this feedback is a positive note!

reneleonhardt committed:
More updates

reneleonhardt commented:
Thank you for trying my PR, feels better than being ignored and eventually closed...
I honestly don't know why this hasn't been merged yet, I heard only positive feedback so far.
But at least the maintainers have 5 months to merge this... before Python 3.14 will be released 😉
Which by the way also worked with my PR, at least as far as the pytest coverage is concerned.
If you want you can pull again, I rebased, resolved conflicts and updated a little more.

Slach commented:
@reneleonhardt, could you resolve conflicts in this branch?

reneleonhardt commented:
This good old pip-compile style dependency management nobody is using anymore for years is an absolute nightmare.
I could only spend hours again if my work would be constructively reviewed in a timely fashion and then merged if there are no problems.
I have no trust in the maintainers anymore after adding the priority label and then keeping me waiting for 5 months.

Updates are literally time-sensitive, now every dependency constraint in every "requirement" file has to be reevaluated to satisfy all supported Python versions (supposedly 3.10, 3.11, 3.12 and maybe 3.13 if my work would've been merged).

It would be much easier to throw all ancient requirement files away and support the packaging standard (groups in pyproject.toml):
https://packaging.python.org/en/latest/specifications/dependency-groups/
https://peps.python.org/pep-0735/
https://github.com/Aider-AI/aider/blob/main/scripts/pip-compile.sh

akostadinov commented:
Seems like this one can already be closed, see #4899 , wtill thank you for your effort, it got me started.
